Delayed/repeat power effects

cyanpill opened the pull request:
Added "delayedEffect", for power effects that start after a timer, and/or repeat
Added a vector of delayedEffects to each entity, which works with players and critters

I was testing logging out/zoning/dying while an effect was ticking down, and found that activation didn't properly check those as well. Oops! So I refactored checks for power range, to catch any null entity. Worked in every variation of player vs player vs critter I could think of.

Minor issue: the effects use the "character power" of the source to determine the strength of the effect, which isn't accessible without the entity. Ideally the strength would be calculated first. For now the effects end early, which was not as it was on live.

broxen commented on Projects/CoX/Servers/MapServer/DataHelpers.cpp:
```diff
@@ -780,18 +780,7 @@ bool checkPowerRange(Entity &ent, Entity &target_ent, float range)
             return false;
     return true;
 }
-bool checkPowerRange(Entity &ent, uint32_t tgt_idx, uint32_t pset_idx, uint32_t pow_idx)
```
Was removing this intentional? Looks like we're still using it in `checkDelayedEffects()`

broxen replied:
Nevermind, I see that this was overloaded.

broxen commented on Projects/CoX/Servers/MapServer/WorldSimulation.cpp:
```diff
@@ -88,10 +88,36 @@ void World::effectsStep(Entity *e,uint32_t msec)
+            ++dly_idx;
+    }
+}
 void World::checkPowerTimers(Entity *e, uint32_t msec)
```
This method has become a monster. In a future PR, can we refactor this into several smaller chunks?
```cpp
checkActivationTimers();
checkRechargeTimers();
checkAutoToggleTimers();
checkBuffTimers();
```
or something similar?

broxen replied:
Maybe we break some of this out into another file as well to keep WorldSimulation.cpp lean and mean.

cyanpill replied:
I agree, originally I had the delay stuff in powertimers, this is much cleaner.

broxen commented:
I pulled this down for some testing and ran into a segfault.

It appears to happen when you turn on auto-attack (right-click power) and target a Critter. While auto attacking continue to attack with other powers. With the delayed powers (charging) There is a small window where the Critter has died and is removed from the map, but the next auto attack fires anyway, resulting in:

```
[00:41:28] [log.slashcommand]Debug   : Sending 17 FloatingNumbers from 263 to 929
// charging text
[00:41:33] [log.slashcommand]Debug   : Sending 28 FloatingNumbers from 263 to 929
[00:41:33] Warning : "Entity 929 does not exist, or is not currently online."
```

cyanpill commented:
Ok, turns out auto-attack was not checking if the target was dead either. So I added a few checks to see if a target is dead, might be redundant in some cases, but it works.
I looked into fade timing(#872) and it turns out we are not actually setting that? So when critters die, I start the fade timer, which seems to take around 15 seconds. They are unable to be affected by most powers during that timer.
Also added a get entity that returns the source if nothing is found(#871). Changed over my checks to this new function.
And split up the checkpowertimers() as discussed.

broxen commented:
You're amazing Cyan 😍

nemerle commented on Projects/CoX/Common/GameData/Powers.h:
```diff
@@ -336,7 +336,14 @@ static const int m_num_trays = 2; // was 3, displayed trays
+    uint32_t src_ent;
+    StoredAttribMod mod;
+    CharacterPower *power;
+};
```
I'd consider a slight reordering of the struct fields, 
```cpp
{
    StoredAttribMod mod;
    CharacterPower *power;
    int m_timer;
    int ticks;
    uint32_t src_ent;
};
```
Unless access patterns say otherwise :)

nemerle commented on Projects/CoX/Servers/MapServer/DataHelpers.cpp:
```diff
@@ -987,7 +976,14 @@ void findAttrib(Entity &ent, Entity *target_ent, CharacterPower * ppower)
+                doAtrrib(ent, target_ent, powtpl.pAttribMod[i], ppower);
+            else
+            {           //queue the power up to fire later or multiple times
+                target_ent->m_delayed.push_back(
```
`emplace_back` instead ?

nemerle commented on Projects/CoX/Servers/MapServer/MapInstance.cpp:
```diff
@@ -2331,7 +2331,8 @@ void MapInstance::on_abort_queued_power(AbortQueuedPower * ev)
 
     // remove last queued power
-    session.m_ent->m_queued_powers.pop_back();
+    session.m_ent->m_queued_powers.end()->m_activation_state = false;
```
You should use session.m_ent->m_queued_powers.back() here + check if there are any entries in `session.m_ent->m_queued_powers` at all

nemerle commented on Projects/CoX/Servers/MapServer/WorldSimulation.cpp:
```diff
@@ -85,16 +85,41 @@ void World::effectsStep(Entity *e,uint32_t msec)
 
-void World::checkPowerTimers(Entity *e, uint32_t msec)
+        if(dly_idx->ticks < 0)
+            e->m_delayed.erase(dly_idx);
```
depending on the underlying type of container, You'd want `dly_idx = e->m_delayed.erase(dly_idx);` here.

nemerle commented on Projects/CoX/Servers/MapServer/WorldSimulation.cpp:
```diff
@@ -270,7 +321,17 @@ void World::updateEntity(Entity *e, const ACE_Time_Value &dT)
+    // for now we only run these on players
+    if(e->m_type == EntType::PLAYER)
+    {
+        checkActivationTimers(e, uint32_t(dT.msec()));
```
I'd refactor the dT.msec() calls to:
```cpp
const uint32_t dt_msec = uint32_t(dT.msec());
```
and then use dt_msec afterwards

cyanpill replied:
how about using sim_frame_time? it's already set to delta.msec()/1000.0f, testing shows it works with the same timing as before. And it's public so we don't need to pass it... unless we are moving these functions out of worldsim.cpp

nemerle replied:
One day I'd love to have ECS like architecture for things like this, so using parameter here might not make as much sense as sim_frame_time, but later on, it will be easier.

Quick note, take a look at the refactor that is coming https://github.com/Segs/Segs/blob/6002c76f78f6d93d72faf364005740a6835e4b16/Projects/CoX/Servers/MapServer/WorldSimulation.cpp#L743
